File: Apps/iOS/Logging/TimestampedLog.swift
import Foundation
import os

struct TimestampedLog: LogProtocol {
    private static let timestampStyle = Date.ISO8601FormatStyle(timeZone: .gmt)
        .year()
        .month()
        .day()
        .dateSeparator(.dash)
        .dateTimeSeparator(.space)
        .time(includingFractionalSeconds: true)

    private let logger: Logger

    init(subsystem: String, category: String) {
        logger = Logger(subsystem: subsystem, category: category)
    }

    func record(_ line: String) {
        let timestamped = stamped(line)
        logger.info("\(timestamped, privacy: .public)")
    }

    func recordSample(_ line: String) {
        let timestamped = stamped(line)
        logger.debug("\(timestamped, privacy: .public)")
    }

    private func stamped(_ line: String) -> String {
        "\(Date.now.formatted(Self.timestampStyle)) \(line)"
    }
}

File: Web/src/logging/timestampedLog.ts
import type { Log } from '../domain/protocols/log.js'

export class TimestampedLog implements Log {
    record(line: string): void {
        console.info(stamped(line))
    }

    recordSample(line: string): void {
        console.debug(stamped(line))
    }
}

function stamped(line: string): string {
    return `${new Date().toISOString().replace('T', ' ').replace('Z', '')} ${line}`
}
